File: 257. Binary Tree Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <sstream>
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector <string> res;
        preorder({}, root, res);
        return res;
    }
    
    void preorder(vector <int> path, TreeNode* root, vector <string> &res)
    {
        if (root == NULL)
            return;
        if (root->left == NULL && root->right == NULL)
        {
            path.push_back(root->val);
            stringstream ss;
            for(size_t i = 0; i < path.size(); ++i)
            {
                if(i != 0)
                    ss << "->";
                ss << path[i];
            }
            string s = ss.str();
            res.push_back(s);
            path.pop_back();
        }
        path.push_back(root->val);
        preorder(path, root->left, res);
        preorder(path, root->right, res);
        path.pop_back();
    }
};

File: 257. Binary Tree Paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {string[]}
    def binaryTreePaths(self, root):
        if not root:
            return []
        self.res = []
        self.preorder([], root)
        return self.res
        
    def preorder(self, path, root):
        if not root:
            return
        if not root.left and not root.right:
            path += [root.val]
            self.res += ["->".join(str(i) for i in path)]
            path.pop()
        path += [root.val]
        self.preorder(path, root.left)
        self.preorder(path, root.right)
        path.pop()
